src/cmap/mod.rs

```diff
@@ -388,37 +295,101 @@ impl ConnectionPoolInner {
-            self.close_connection(conn, ConnectionClosedReason::Stale);
+        let mut connection_manager = self.connection_manager.lock().await;
 
-            return;
+        // Close the connection if it's stale.
+        if conn.is_stale(connection_manager.generation) {
+            connection_manager.close_connection(conn, ConnectionClosedReason::Stale);
+        } else {
+            connection_manager.checked_in_connections.push(conn);
         }
 
-        self.connections.write().unwrap().push(conn);
-        self.wait_queue.notify_ready();
+        self.wait_queue.wake_front();
     }
 
-    /// Internal helper to close a connection, emit the event for it being closed, and decrement the
-    /// total connection count. Any connection being closed by the pool should be closed by using
-    /// this method.
-    fn close_connection(&self, conn: Connection, reason: ConnectionClosedReason) {
+    async fn check_out(&self) -> Result<Connection> {
         self.emit_event(|handler| {
-            handler.handle_connection_closed_event(conn.closed_event(reason));
+            let event = ConnectionCheckoutStartedEvent {
+                address: self.address.clone(),
+            };
+
+            handler.handle_connection_checkout_started_event(event);
         });
 
-        self.total_connection_count.fetch_sub(1, Ordering::SeqCst);
+        let result = self.acquire_or_create_connection().await;
+
+        let conn = match result {
+            Ok(conn) => conn,
+            Err(e) => {
+                let failure_reason = if let ErrorKind::WaitQueueTimeoutError { .. } = e.kind.as_ref() {
+                    ConnectionCheckoutFailedReason::Timeout
+                } else {
+                    ConnectionCheckoutFailedReason::ConnectionError
+                };
+
+                self.emit_event(|handler| {
+                    handler.handle_connection_checkout_failed_event(
+                        ConnectionCheckoutFailedEvent {
+                            address: self.address.clone(),
+                            reason: failure_reason,
+                        },
+                    )
+                });
+
+                return Err(e);
+            }
+        };
+
+        self.emit_event(|handler| {
+            handler.handle_connection_checked_out_event(conn.checked_out_event());
+        });
+
+        Ok(conn)
+    }
+
+    /// Waits for the thread to reach the front of the wait queue, then attempts to check out a
+    /// connection. If none are available in the pool, one is created and checked out instead.
+    async fn acquire_or_create_connection(&self) -> Result<Connection> {
+        // Handle that will wake up the front of the queue when dropped.
+        // Before returning a valid connection, this handle must be disarmed to prevent the front
+        // from waking up early.
+        let mut wait_queue_handle = self.wait_queue.wait_until_at_front().await?;
```

Comment:
this handle allows us to use the `?` operator, which is nice. It does need to be disarmed when successfully returning though.